File: Algorithm/swea/swea_4038.py
def KMP_table(pattern):
    lp = len(pattern)
    table = [0]*lp # 정보 저장용 테이블
    
    pidx = 0 # 테이블의 값을 불러오고, 패턴의 인덱스에 접근
    for idx in range(1, lp): # 테이블에 값 저장하기 위해 활용하는 인덱스
        # pidx가 0이 되거나, idx와 pidx의 pattern 접근 값이 같아질때까지 진행
        while pidx and pattern[idx] != pattern[pidx]:
            pidx = table[pidx-1]
        
        # 값이 일치하는 경우, pidx 1 증가시키고 그 값을 tb에 저장
        if pattern[idx] == pattern[pidx] :
            pidx += 1
            table[idx] = pidx
    return table
    
    
def KMP(word, pattern):
    # KMP_table 통해 전처리된 테이블 불러오기
    table = KMP_table(pattern)
    cnt = 0     
    pidx = 0
    
    for idx in range(len(word)):
        # 단어와 패턴이 일치하지 않는 경우, pidx를 table을 활용해 값 변경
        while pidx > 0 and word[idx] != pattern[pidx] :
            pidx = table[pidx-1]
        # 해당 인덱스에서 값이 일치한다면, pidx를 1 증가시킴
        # 만약 pidx가 패턴의 끝까지 도달하였다면, 시작 인덱스 값을 계산하여 추가 후 pidx 값 table의 인덱스에 접근하여 변경
        if word[idx] == pattern[pidx]:
            if pidx == len(pattern)-1 :
                cnt += 1
                pidx = table[pidx]
            else:
                pidx += 1
    
    return cnt

for tc in range(1,int(input())+1):

    a = input()
    b = input()
    print(f"#{tc} {KMP(a,b)}")

File: Algorithm/baekjoon/java/src/swea_4038.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class swea_4038 {
    static int[] table ;

    static void KMP_table(String pattern) {
        int pidx = 0;
        for (int idx=1; idx<pattern.length(); idx++) {
            while (pidx>0 && pattern.charAt(idx) != pattern.charAt(pidx)) {
                pidx = table[pidx-1];
            }
            if (pattern.charAt(idx) == pattern.charAt(pidx)) {
                pidx ++;
                table[idx] = pidx;
            }
        }

    }

    static int KMP(String word, String pattern) {
        KMP_table(pattern);
        int cnt = 0;
        int pidx = 0;

        for (int idx=0; idx<word.length(); idx++) {
            while (pidx>0 && word.charAt(idx) != pattern.charAt(pidx)) {
                pidx = table[pidx-1];
            }

            if (word.charAt(idx) == pattern.charAt(pidx)) {
                if (pidx == pattern.length()-1) {
                    cnt ++;
                    pidx = table[pidx];
                } else {
                    pidx ++ ;
                }
            }
        }
        return cnt ;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        // tc
        StringTokenizer st = new StringTokenizer(br.readLine());
        int TC = Integer.parseInt(st.nextToken());

        for (int tc=1; tc<=TC; tc++) {
            String a = new StringTokenizer(br.readLine()).nextToken();
            String b = new StringTokenizer(br.readLine()).nextToken();
            table = new int[a.length()];
            System.out.println("#"+tc+" "+KMP(a,b));
        }
    }
}
